internal/provider/resource_mackerel_notification_group.go

```diff
@@ -0,0 +1,202 @@
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringdefault"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/mackerelio-labs/terraform-provider-mackerel/internal/mackerel"
+)
+
+var (
+	_ resource.Resource                = (*mackerelNotificationGroupResource)(nil)
+	_ resource.ResourceWithConfigure   = (*mackerelNotificationGroupResource)(nil)
+	_ resource.ResourceWithImportState = (*mackerelNotificationGroupResource)(nil)
+)
+
+func NewMackerelNotificationGroupResource() resource.Resource {
+	return &mackerelNotificationGroupResource{}
+}
+
+type mackerelNotificationGroupResource struct {
+	Client *mackerel.Client
+}
+
+func (r *mackerelNotificationGroupResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_notification_group"
+}
+
+func (r *mackerelNotificationGroupResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: "This resouce allows creating and management of notification groups",
+
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Description: "The ID of the notitication group",
+
+				Computed: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"name": schema.StringAttribute{
+				Description: "The name of the notification group",
+
+				Required: true,
+			},
+			"notification_level": schema.StringAttribute{
+				MarkdownDescription: "The level of notitication (`all` or `critical`)",
+
+				Optional: true,
+				Computed: true,
+				Validators: []validator.String{
+					mackerel.NotificationLevelValidator(),
+				},
+				Default: stringdefault.StaticString("all"),
+			},
+			"child_notification_group_ids": schema.SetAttribute{
+				Description: "A set of notification group IDs",
+
+				ElementType: types.StringType,
+				Optional:    true,
+			},
+			"child_channel_ids": schema.SetAttribute{
+				Description: "A set of notification channel IDs",
+
+				ElementType: types.StringType,
+				Optional:    true,
+			},
+		},
+		// TODO: migrate to nested attributes
+		Blocks: map[string]schema.Block{
+			"monitor": schema.SetNestedBlock{
+				Description: "Configuration block(s) with monitor rules",
+
+				NestedObject: schema.NestedBlockObject{
+					Attributes: map[string]schema.Attribute{
+						"id": schema.StringAttribute{
+							Description: "The monitor rule ID",
+
+							Required: true,
+						},
+						"skip_default": schema.BoolAttribute{
+							Description: "If true, send notifications to this notification group only.",
+
+							Optional: true,
+							Computed: true,
+							Default:  booldefault.StaticBool(false),
```

Comment:
Attributes with default value are both inputs and outputs. If the user provides a value, they are act as input. If not, the provider computes and provides a default value. So they need not to be only optional but also computed.